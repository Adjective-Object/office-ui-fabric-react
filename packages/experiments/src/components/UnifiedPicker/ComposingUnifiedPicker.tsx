import * as React from 'react';
import { DefaultPickerFooterItems } from './DefaultPickerFooterItems';
import { UnifiedPickerSelectedItemsProps, UnifiedPickerFloatingPickerProps } from './UnifiedPicker.types';
import { FloatingSuggestions, SuggestionsStore, IFloatingSuggestionsInnerSuggestionProps } from './../FloatingSuggestions';
import { SelectedItemsList } from '../SelectedItemsList';
import { SuggestionsControl } from '../FloatingSuggestions';
import { UnifiedPicker, UnifiedPickerImpl } from '../UnifiedPicker';
import {
  ComposingUnifiedPickerProps,
  ComposableSuggestionControl,
  ComposableMainFloatingPicker,
  PropsOf
} from './ComposingUnifiedPicker.types';

/**
 * Attep
 */
export class ComposingUnifiedPicker<T> extends React.PureComponent<ComposingUnifiedPickerProps<T>> {
  private _picker: UnifiedPickerImpl<T>;
  // Custom footer items that reflect the state of the picker.
  private _defaultFooterItems: DefaultPickerFooterItems = new DefaultPickerFooterItems();

  constructor(props: ComposingUnifiedPickerProps<T>) {
    super(props);
  }

  public render(): JSX.Element {
    return (
      <UnifiedPicker
        onRenderFloatingPicker={this.MainFloatingPicker}
        onRenderSelectedItems={this.DefaultSelectedItemsList}
        onRenderFocusZone={this.props.onRenderFocusZone}
        key={'normal'}
        inputProps={{
          'aria-label': 'People Picker'
        }}
        componentRef={this._setComponentRef}
        // Styling props pasthrough
        className={this.props.className}
        styles={this.props.styles}
        disabled={this.props.disabled}
      />
    );
  }

  /**
   * the default selected items list
   */
  private DefaultSuggestionControlInner: ComposableSuggestionControl<T> = (overriddenProps: PropsOf<ComposableSuggestionControl<T>>) => (
    <SuggestionsControl
      showRemoveButtons={this._shouldShowSuggestionRemoveButtons()}
      headerItemsProps={[]}
      footerItemsProps={this._defaultFooterItems.items}
      shouldSelectFirstItem={this._inputIsNotEmpty}
      {...overriddenProps}
    />
  );

  /**
   * Renders the suggestion control that is passed to both the
   * main picker and the editing item picker.
   *
   * Uses the default suggestion control if none is provided.
   */
  private SuggestionControl = (overriddenProps: IFloatingSuggestionsInnerSuggestionProps<T>) => {
    const Inner: ComposableSuggestionControl<T> = this.props.onRenderSuggestionControl || this.DefaultSuggestionControlInner;
    return <Inner onRenderSuggestion={this.props.onRenderSuggestionItem} {...overriddenProps} />;
  };

  /**
   * The floating picker that is used by default for both the
   * floating picker suggestion dropdown and the suggestion well floating picker.
   */
  private DefaultFloatingPickerInner: ComposableMainFloatingPicker<T> = (overriddenProps: PropsOf<ComposableMainFloatingPicker<T>>) => (
    <FloatingSuggestions
      suggestionsStore={new SuggestionsStore<T>()}
      onResolveSuggestions={this.props.onResolveSuggestions}
      onRemoveSuggestion={this.props.onRemoveSuggestion}
      onZeroQuerySuggestion={this.props.onZeroQuerySuggestion}
      isQueryForceResolveable={this.props.isQueryForceResolveable}
      onRenderSuggestionsItem={this.props.onRenderSuggestionItem}
      {...overriddenProps}
    />
  );

  /**
   * Renders the floating picker for the main input.
   * Uses the default picker if none is provided.
   */
  private MainFloatingPicker = (overriddenProps: UnifiedPickerFloatingPickerProps<T>) => {
    const Inner: ComposableMainFloatingPicker<T> = this.props.onRenderMainFloatingPicker || this.DefaultFloatingPickerInner;
    return <Inner onRenderSuggestionControl={this.SuggestionControl} {...overriddenProps} />;
  };

  /**
   * the default selected items list
   */
  private DefaultSelectedItemsList = (overriddenProps: UnifiedPickerSelectedItemsProps<T>) => (
    <SelectedItemsList onRenderItem={this.props.onRenderSelectedItem} {...overriddenProps} />
  );

  private _setComponentRef = (component: UnifiedPickerImpl<T>): void => {
    this._picker = component;
    if (this.props.onRenderSuggestionControl === undefined) {
      // Only bind the default footer items to the component
      // if we are using the default suggestions control.
      this._defaultFooterItems.bindPickerInstance(this._picker);
    } else {
      // Clean up the reference otherwise
      this._defaultFooterItems.bindPickerInstance(undefined);
    }
  };

  private _inputIsNotEmpty = (): boolean => {
    return !(this._picker !== undefined && this._picker.inputElement !== null && this._picker.inputElement.value === '');
  };

  /**
   * If the suggestion remove buttons should be shown or nots
   *
   * Show if we control the data model
   */
  private _shouldShowSuggestionRemoveButtons() {
    return this.props.onRemoveSuggestion !== undefined;
  }
}
