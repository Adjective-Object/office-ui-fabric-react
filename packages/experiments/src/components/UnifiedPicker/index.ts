export * from './UnifiedPicker';
export * from './UnifiedPicker.types';
