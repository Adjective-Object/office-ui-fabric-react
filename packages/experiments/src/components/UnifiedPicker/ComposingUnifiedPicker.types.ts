import * as React from 'react';
import { IFloatingSuggestionsInnerSuggestionProps, IFloatingSuggestionsProps } from '../FloatingSuggestions/FloatingSuggestions.types';
import { ISuggestionsControlProps } from '../FloatingSuggestions/Suggestions/Suggestions.types';
import { ISelectedItemsListProps } from '../SelectedItemsList';
import { IUnifiedPickerProps, IUnifiedPickerComponentPassthroughProps } from './UnifiedPicker.types';

export type PropsOf<T> = T extends React.ComponentType<infer P> ? P : never;
export type WithMoreProps<TComponent extends React.ComponentType, TMoreProps> = React.ComponentType<PropsOf<TComponent> & TMoreProps>;

/**
 * Expand the exact floating picker props that are passed through from the UnifiedPicker
 * so that if FloatingPicker is overridden, the consumer does not need to override the full
 * component subtree (e.g. the separately specified or default component)
 */
export type ComposableMainFloatingPicker<T> = WithMoreProps<
  IUnifiedPickerProps<T>['onRenderFloatingPicker'],
  Pick<IFloatingSuggestionsProps<T>, 'onRenderSuggestionControl'>
>;

/**
 * Expand the exact suggestion control props that are passed through from the FloatingPicker
 * so that if SuggestionControl is overridden, the consumer does not need to override the full
 * component subtree (e.g. the separately specified or default component)
 */
export type ComposableSuggestionControl<T> = React.ComponentType<
  IFloatingSuggestionsInnerSuggestionProps<T> & Pick<ISuggestionsControlProps<T>, 'onRenderSuggestion'>
>;

export type ComposingUnifiedPickerProps<T> = {
  /**
   * Passed down to UnifiedPeoplePicker div. Required for fabric compliance.
   */
  className?: string;

  /////////////////////////////////////
  // Coordinated Internal Components //
  /////////////////////////////////////
  onRenderFocusZone?: IUnifiedPickerProps<T>['onRenderFocusZone'];
  onRenderMainFloatingPicker?: ComposableMainFloatingPicker<T>;
  onRenderSuggestionControl?: ComposableSuggestionControl<T>;
  onRenderSuggestionItem: ISuggestionsControlProps<T>['onRenderSuggestion'];
  onRenderSelectedItem: ISelectedItemsListProps<T>['onRenderItem'];

  ///////////////////////////
  // Data Model (required) //
  ///////////////////////////
  onResolveSuggestions: IFloatingSuggestionsProps<T>['onResolveSuggestions'];

  //////////////////////////////////////
  // Customizable Behavior (optional) //
  //////////////////////////////////////
  isQueryForceResolveable?: IFloatingSuggestionsProps<T>['isQueryForceResolveable'];
  onZeroQuerySuggestion?: IFloatingSuggestionsProps<T>['onZeroQuerySuggestion'];

  /////////////////////////////////////////////
  // Props for use as a controlled component //
  // TODO: write a wrapper component that    //
  // manages state for use as uncontrolled   //
  /////////////////////////////////////////////
  onRemoveSuggestion?: IFloatingSuggestionsProps<T>['onRemoveSuggestion'];
} & IUnifiedPickerComponentPassthroughProps;
