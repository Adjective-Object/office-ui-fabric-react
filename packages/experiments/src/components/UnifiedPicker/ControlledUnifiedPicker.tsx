import * as React from 'react';
import { KeyCodes, classNamesFunction } from 'office-ui-fabric-react/lib/Utilities';
import { Autofill } from 'office-ui-fabric-react/lib/Autofill';
import { IInputProps } from 'office-ui-fabric-react/lib/Pickers';
import { IFloatingSuggestions } from '../FloatingSuggestions';
import { IControlledSelectedItemsList } from '../SelectedItemsList';
import { FocusZone, FocusZoneDirection, FocusZoneTabbableElements } from 'office-ui-fabric-react/lib/FocusZone';
import { Selection, SelectionMode, SelectionZone } from 'office-ui-fabric-react/lib/Selection';

import { getStyles } from './UnifiedPicker.styles';
import { styled } from 'office-ui-fabric-react/lib/Utilities';

import {
  UnifiedPickerFocusZoneProps,
  IUnifiedPickerStyleProps,
  IUnifiedPickerStyles,
  IControlledUnifiedPickerProps,
  IControlledUnifiedPicker
} from './UnifiedPicker.types';

/**
 * The focus zone that is used by default to wrap the whole unified picker
 */
const DefaultUnifiedPickerFocusZone = (overriddenProps: UnifiedPickerFocusZoneProps) => (
  <FocusZone
    shouldInputLoseFocusOnArrowKey={shouldFocusZoneInputLoseFocusOnArrowKey}
    handleTabKey={FocusZoneTabbableElements.all}
    {...overriddenProps}
  />
);
const shouldFocusZoneInputLoseFocusOnArrowKey = () => true;

const getClassNames = classNamesFunction<IUnifiedPickerStyleProps, IUnifiedPickerStyles>();

/**
 * Internal controlled component for UnifiedPicker.
 *
 * Handles UI state for the picker. Data model is controlled by its parent, UncontrolledUnifiedPicker.
 *
 * Hanldes rendering and hooking up UI.
 * @param props
 */
export const ControlledUnifiedPicker = React.forwardRef(
  styled(
    <TSelectedItem extends any, TSuggestedItem = TSelectedItem>(
      props: IControlledUnifiedPickerProps<TSelectedItem, TSuggestedItem>,
      ref: React.Ref<IControlledUnifiedPicker<TSelectedItem, TSuggestedItem>>
    ) => {
      const {
        className,
        styles,
        theme,
        disabled,
        queryString,
        selectedItems,
        canAddItems,
        onSelectedItemsRemoved,
        onQueryStringChange,
        onPaste,
        onSuggestionSelected,
        isQueryForceResolveable,
        inputProps,
        onRenderFocusZone,
        onRenderSelectedItems,
        onRenderFloatingSuggestions
      } = props;
      const classNames = getClassNames(styles, {
        theme: theme!,
        disabled,
        className
      });

      const floatingSuggestionsRef = React.useRef<IFloatingSuggestions<TSelectedItem>>(null);
      const autofillRef = React.useRef<Autofill>(null);
      const selectedItemsRef = React.useRef<IControlledSelectedItemsList>(null);

      const selection = React.useMemo(() => new Selection(), []);

      /**
       * Calls parent onBackspace behaviour when backspace is pressed in the focus zone
       */
      const removeLastSelectedElementOnEmptyInputBackspace = React.useCallback(
        (ev: React.KeyboardEvent<HTMLElement>) => {
          if (ev.which === KeyCodes.backspace && autofillRef.current && autofillRef.current.value === '') {
            if (selectedItems.length) {
              onSelectedItemsRemoved([selectedItems[selectedItems.length - 1]]);
            }

            if (floatingSuggestionsRef.current) {
              floatingSuggestionsRef.current.hidePicker();
            }
          }
        },
        [props.onSelectedItemsRemoved]
      );

      /**
       * Show the suggestions and clear the selected items in the well when the input is clicked
       */
      const clearSelectionAndShowSuggestions = React.useCallback(() => {
        selection.setAllSelected(false);
        floatingSuggestionsRef.current && floatingSuggestionsRef.current.showPicker();
      }, [floatingSuggestionsRef, selection]);

      const copySelectedItems = React.useCallback(() => {
        selectedItemsRef.current && selectedItemsRef.current.copyItemsInSelectionToClipboard();
      }, [selectedItemsRef]);

      const onAutofillUpdated = React.useCallback(
        (ev: React.ChangeEvent<HTMLInputElement>) => {
          onQueryStringChange(ev.target.value);
        },
        [onQueryStringChange]
      );

      /**
       * Calculates the global ID of the suggestion that is "active" right now.
       *
       * note: depends on only one suggestion being open at a time on the page.
       * We may need to amend SuggestionStore to support prefixed ID spaces for
       * activeDescendant.
       */
      const activeDescendant = React.useMemo(
        () =>
          floatingSuggestionsRef.current && floatingSuggestionsRef.current.currentSelectedSuggestionIndex !== -1
            ? 'sug-' + floatingSuggestionsRef.current.currentSelectedSuggestionIndex
            : undefined,
        [floatingSuggestionsRef, floatingSuggestionsRef.current && floatingSuggestionsRef.current.currentSelectedSuggestionIndex]
      );

      /**
       * Imperative handle to this component
       */
      React.useImperativeHandle(
        ref,
        () => ({
          focusInput: () => {
            autofillRef.current && autofillRef.current.focus();
          }
        }),
        [autofillRef]
      );

      const FocusZoneComponent = onRenderFocusZone || DefaultUnifiedPickerFocusZone;
      const FloatingSuggestions = onRenderFloatingSuggestions;
      const SelectedItems = onRenderSelectedItems;

      return (
        <>
          <FocusZoneComponent
            className={classNames.root}
            direction={FocusZoneDirection.bidirectional}
            onKeyDown={removeLastSelectedElementOnEmptyInputBackspace}
            onCopy={copySelectedItems}
          >
            <SelectionZone selection={selection} selectionMode={SelectionMode.multiple}>
              <div className={classNames.pickerWell} role={'list'}>
                <SelectedItems selection={selection} selectedItems={selectedItems} onItemsRemoved={onSelectedItemsRemoved} />
                {canAddItems && (
                  <Autofill
                    {...inputProps as IInputProps}
                    ref={autofillRef}
                    className={classNames.input}
                    value={queryString}
                    onChange={onAutofillUpdated}
                    onFocus={clearSelectionAndShowSuggestions}
                    onClick={clearSelectionAndShowSuggestions}
                    onInputValueChange={onQueryStringChange}
                    aria-activedescendant={activeDescendant}
                    aria-owns="suggestion-list"
                    aria-expanded={floatingSuggestionsRef.current ? floatingSuggestionsRef.current.isSuggestionsShown : false}
                    aria-haspopup="true"
                    autoCapitalize="off"
                    autoComplete="off"
                    role="combobox"
                    disabled={disabled}
                    onPaste={onPaste}
                  />
                )}
              </div>
            </SelectionZone>
          </FocusZoneComponent>
          {canAddItems && autofillRef && autofillRef.current && (
            <FloatingSuggestions
              ref={floatingSuggestionsRef}
              inputElement={autofillRef.current.inputElement}
              onSuggestionSelected={onSuggestionSelected}
              isQueryForceResolveable={isQueryForceResolveable}
            />
          )}
        </>
      );
    },
    getStyles,
    undefined
  )
  // Cast back to a generic function type, since typescript <3.4 collapses generics on higher-order functions
  // This was solved by https://github.com/microsoft/TypeScript/pull/30215 in typescript@3.4,
) as <TSelectedItem extends any, TSuggestedItem = TSelectedItem>(
  props: IControlledUnifiedPickerProps<TSelectedItem, TSuggestedItem> &
    React.RefAttributes<IControlledUnifiedPicker<TSelectedItem, TSuggestedItem>>
) => React.ReactElement;
export type ControlledUnifiedPicker<TSelectedItem, TSuggestedItem> = (
  props: IControlledUnifiedPickerProps<TSelectedItem, TSuggestedItem> &
    React.RefAttributes<IControlledUnifiedPicker<TSelectedItem, TSuggestedItem>>
) => React.ReactElement;
