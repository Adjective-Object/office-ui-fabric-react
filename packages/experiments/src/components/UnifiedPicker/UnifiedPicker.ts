import { UnifiedPickerView } from './UnifiedPicker.view';
import { UnifiedPickerStyles, UnifiedPickerTokens } from './UnifiedPicker.styles';
import { UnifiedPickerState } from './UnifiedPicker.state';
import { IUnifiedPickerProps } from './UnifiedPicker.types';
import { createComponent } from '../../Foundation';

export const UnifiedPicker: React.StatelessComponent<IUnifiedPickerProps> = createComponent({
  displayName: 'UnifiedPicker',
  view: UnifiedPickerView,
  state: UnifiedPickerState,
  styles: UnifiedPickerStyles,
  tokens: UnifiedPickerTokens
});
