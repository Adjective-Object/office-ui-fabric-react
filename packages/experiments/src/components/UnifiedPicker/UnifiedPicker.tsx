import * as React from 'react';

import { IUnifiedPickerProps, IUnifiedPicker } from './UnifiedPicker.types';

function isPromiseLike<T>(r: T | PromiseLike<T>): r is PromiseLike<T> {
  return !!(r && (r as { then?: Function }).then);
}

const maybeAwait = <T extends any>(inputValue: T | PromiseLike<T>): PromiseLike<T> => {
  if (isPromiseLike(inputValue)) {
    return inputValue;
  }
  return {
    then: (callback: (value: T) => any) => callback(inputValue)
  };
};

/**
 * A state-only component that manages a UnifiedPicker's internal state
 *
 * (e.g. managing a list of items against consumer-provided hooks)
 *
 * Renders the view that is passed into it.
 */
export const UnifiedPicker = React.forwardRef(
  <TSelectedItem, TSuggestedItem = TSelectedItem, TViewRefTarget = any>(
    props: IUnifiedPickerProps<TSelectedItem, TSuggestedItem>,
    ref: React.Ref<IUnifiedPicker<TSelectedItem, TSuggestedItem, TViewRefTarget>>
  ) => {
    const { defaultSelectedItems, defaultQueryString, processSelectedItem, createItemFromPasteData } = props;

    const [selectedItems, setSelectedItems] = React.useState(defaultSelectedItems || []);
    const [queryString, updateQueryString] = React.useState(defaultQueryString || '');

    const onSelectedItemsRemoved = React.useCallback(
      (removedItems: TSelectedItem[]) => {
        console.log('unifiedPicker onSelectedItemsRemoved called', removedItems);
        // Update internal state if it is tracked at all
        const nextSelectedItems = selectedItems.filter(i => removedItems.indexOf(i) === -1);
        if (nextSelectedItems.length !== selectedItems.length - removedItems.length) {
          console.error('UnifiedPicker onSelectedItemsRemoved called on some item not in the current state');
        }
        setSelectedItems(nextSelectedItems);
      },
      [selectedItems]
    );

    const onSelectedItemChanged = React.useCallback(
      (newItem: TSelectedItem | TSelectedItem[], index: number): void => {
        const newItemsArray = !Array.isArray(newItem) ? [newItem] : newItem;

        if (index >= 0) {
          const newItems: TSelectedItem[] = [...selectedItems];
          newItems.splice(index, 1, ...newItemsArray);
          setSelectedItems(newItems);
        }
      },
      [setSelectedItems, selectedItems]
    );
    const onSuggestionSelected = React.useCallback(
      (item: TSuggestedItem): void => {
        const initialQueryString = queryString;
        const processedItem: TSelectedItem | PromiseLike<TSelectedItem> | null = processSelectedItem
          ? processSelectedItem(item)
          : // procesSelectedItem is only optional if
            // TsuggestedItem is assignable to TSuggestedItem
            ((item as unknown) as TSelectedItem);

        processedItem &&
          maybeAwait(processedItem).then(newItem => {
            // Only actually select the item if the query string hasn't changed since
            // we started resolving the item.
            //
            // This is to handle the case where a user types quickly and presses enter
            // before updating the query string
            if (initialQueryString === queryString) {
              setSelectedItems(selectedItems.concat([newItem]));
            }
          });
        if (processedItem === null) {
          return;
        }
      },
      [queryString, processSelectedItem, selectedItems, setSelectedItems]
    );

    const onPaste = React.useCallback(
      (event: React.ClipboardEvent<any>): void => {
        if (createItemFromPasteData) {
          event.preventDefault();

          const pastedItem: TSelectedItem | PromiseLike<TSelectedItem> = createItemFromPasteData(event.clipboardData);
          pastedItem &&
            maybeAwait(pastedItem).then(newItem => {
              setSelectedItems(selectedItems.concat([newItem]));
            });
        }
      },
      [createItemFromPasteData, selectedItems, setSelectedItems]
    );

    const viewRef = React.useRef<TViewRefTarget>(null);

    // Imperative Handle
    React.useImperativeHandle(
      ref,
      () => ({
        view: viewRef,
        get value() {
          return selectedItems;
        },
        get queryString() {
          return queryString;
        }
      }),
      [viewRef, selectedItems, queryString]
    );

    const View = props.view;
    return (
      <View
        componentRef={viewRef}
        // state
        selectedItems={selectedItems}
        queryString={queryString}
        canAddItems={props.itemLimit === undefined || selectedItems.length < props.itemLimit}
        // state manipulation callbacks
        onSelectedItemsRemoved={onSelectedItemsRemoved}
        onSelectedItemChanged={onSelectedItemChanged}
        onQueryStringChange={updateQueryString}
        onPaste={onPaste}
        onSuggestionSelected={onSuggestionSelected}
      />
    );
  }
  // Cast back to a generic function type, since typescript <3.4 collapses generics on higher-order functions
  // This was solved by https://github.com/microsoft/TypeScript/pull/30215 in typescript@3.4,
) as <TSelectedItem, TSuggestedItem = TSelectedItem, TViewRefTarget = any>(
  props: IUnifiedPickerProps<TSelectedItem, TSuggestedItem> &
    React.RefAttributes<IUnifiedPicker<TSelectedItem, TSuggestedItem, TViewRefTarget>>
) => React.ReactElement;
export type UnifiedPicker<TSelectedItem, TSuggestedItem = TSelectedItem, TViewRefTarget = any> = (
  props: IUnifiedPickerProps<TSelectedItem, TSuggestedItem> &
    React.RefAttributes<IUnifiedPicker<TSelectedItem, TSuggestedItem, TViewRefTarget>>
) => React.ReactElement;
(UnifiedPicker as any).displayName = 'UnifiedPicker';
