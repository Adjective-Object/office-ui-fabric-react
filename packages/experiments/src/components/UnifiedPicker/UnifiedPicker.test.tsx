/* tslint:disable:no-unused-variable */
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';
/* tslint:enable:no-unused-variable */
import * as renderer from 'react-test-renderer';

import { IUnifiedPickerProps, UnifiedPickerFloatingPickerProps } from './UnifiedPicker.types';
import { UnifiedPickerImpl } from './UnifiedPicker';
import { FloatingSuggestions, SuggestionsStore } from '../FloatingSuggestions';
import { ISelectedItemProps, ISelectedItemsListProps } from '../SelectedItemsList/SelectedItemsList.types';
import { SelectedItemsList } from '../SelectedItemsList/SelectedItemsList';
import { KeyCodes } from 'office-ui-fabric-react/lib/Utilities';
import { ISuggestionModel } from 'office-ui-fabric-react/lib/Pickers';

function onResolveSuggestions(text: string): ISimple[] {
  return [
    'black',
    'blue',
    'brown',
    'cyan',
    'green',
    'magenta',
    'mauve',
    'orange',
    'pink',
    'purple',
    'red',
    'rose',
    'violet',
    'white',
    'yellow'
  ]
    .filter((tag: string) => tag.toLowerCase().indexOf(text.toLowerCase()) === 0)
    .map((item: string) => ({ key: item, name: item }));
}

const basicSuggestionRenderer = (props: ISuggestionModel<ISimple>) => {
  return <div key={props.item.key}> {props.item.name} </div>;
};

const basicItemRenderer = (props: ISelectedItemProps<ISimple>) => {
  return <div key={props.key}> {props.name} </div>;
};

const BasicFloatingSuggestions = <T extends UnifiedPickerFloatingPickerProps<ISimple>>(props: T) => {
  return (
    <FloatingSuggestions
      onResolveSuggestions={onResolveSuggestions}
      onRenderSuggestionsItem={basicSuggestionRenderer}
      suggestionsStore={new SuggestionsStore<ISimple>()}
      {...props}
    />
  );
};

const BasicSelectedItemsList = (props: ISelectedItemsListProps<ISimple>) => {
  return <SelectedItemsList<ISimple> onRenderItem={basicItemRenderer} {...props} />;
};

export interface ISimple {
  key: string;
  name: string;
}

export type TypedBaseUnifiedPicker = UnifiedPickerImpl<ISimple>;

describe('UnifiedPicker', () => {
  const BaseUnifiedPickerWithType: new (props: IUnifiedPickerProps<ISimple>) => UnifiedPickerImpl<ISimple> = UnifiedPickerImpl;

  it('renders BaseUnifiedPicker correctly', () => {
    const component = renderer.create(
      <BaseUnifiedPickerWithType onRenderSelectedItems={BasicSelectedItemsList} onRenderFloatingPicker={BasicFloatingSuggestions} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('force resolves to the first suggestion', () => {
    const root = document.createElement('div');
    document.body.appendChild(root);

    const overrideOnResolveSuggestions = () => [{ key: 'a', name: 'test1' }, { key: 'b', name: 'test2' }];

    const renderSuggestions = (props: UnifiedPickerFloatingPickerProps<ISimple>) => (
      <BasicFloatingSuggestions {...props} onResolveSuggestions={overrideOnResolveSuggestions} />
    );

    const picker: TypedBaseUnifiedPicker = (ReactDOM.render(
      <BaseUnifiedPickerWithType onRenderSelectedItems={BasicSelectedItemsList} onRenderFloatingPicker={renderSuggestions} />,
      root
    ) as unknown) as TypedBaseUnifiedPicker;

    expect(picker.inputElement).toBeTruthy();
    if (picker.inputElement) {
      picker.inputElement.value = 'bl';
      picker.inputElement.focus();
    }

    expect(picker.floatingPicker.current && picker.floatingPicker.current.suggestions.length).toBe(2);
    expect(picker.floatingPicker.current && picker.floatingPicker.current.suggestions[0].item.name).toBe('test1');

    // Force resolve to the first suggestions
    expect(picker.floatingPicker.current).toBeTruthy();
    picker.floatingPicker.current && picker.floatingPicker.current.forceResolveSuggestion();

    expect(picker.value.length).toBe(1);
    expect(picker.value[0].name).toBe('test1');

    ReactDOM.unmountComponentAtNode(root);
  });

  it('Completes the suggestion', () => {
    const root = document.createElement('div');
    document.body.appendChild(root);

    const picker: TypedBaseUnifiedPicker = (ReactDOM.render(
      <BaseUnifiedPickerWithType onRenderSelectedItems={BasicSelectedItemsList} onRenderFloatingPicker={BasicFloatingSuggestions} />,
      root
    ) as unknown) as TypedBaseUnifiedPicker;

    if (picker.inputElement) {
      picker.inputElement.value = 'bl';
      ReactTestUtils.Simulate.keyDown(picker.inputElement, { which: KeyCodes.down });
    }

    picker.floatingPicker.current && picker.floatingPicker.current.onCurrentlySelectedSuggestionChosen();
    expect(picker.selectedItemsList.current && picker.value.length).toBe(1);
    expect(picker.selectedItemsList.current && picker.value[0].name).toBe('blue');

    ReactDOM.unmountComponentAtNode(root);
  });

  describe('As an uncontrolled component going through a state change', () => {
    it('Does not remove elements when pressing backspace over a non-empty input', () => {
      const root = document.createElement('div');
      document.body.appendChild(root);
      const onItemsRemovedSpy = jasmine.createSpy();

      const picker: TypedBaseUnifiedPicker = (ReactDOM.render(
        <BaseUnifiedPickerWithType
          defaultSelectedItems={[{ key: '1', name: 'one' }, { key: '2', name: 'two' }]}
          onRenderSelectedItems={BasicSelectedItemsList}
          onRenderFloatingPicker={BasicFloatingSuggestions}
          onItemsRemoved={onItemsRemovedSpy}
        />,
        root
      ) as unknown) as TypedBaseUnifiedPicker;

      if (!picker.inputElement) {
        throw new Error('picker input element not set!');
      }
      picker.inputElement.value = 'hi';
      ReactTestUtils.Simulate.keyDown(picker.inputElement, { which: KeyCodes.backspace });

      expect(onItemsRemovedSpy).not.toBeCalled();
      expect(picker.value.length).toBe(2);

      ReactDOM.unmountComponentAtNode(root);
    });

    it('Removes the last element when pressing backspace over an empty input', () => {
      const root = document.createElement('div');
      document.body.appendChild(root);
      const onItemsRemovedSpy = jasmine.createSpy();

      const picker: TypedBaseUnifiedPicker = (ReactDOM.render(
        <BaseUnifiedPickerWithType
          defaultSelectedItems={[{ key: '1', name: 'one' }, { key: '2', name: 'two' }]}
          onRenderSelectedItems={BasicSelectedItemsList}
          onRenderFloatingPicker={BasicFloatingSuggestions}
          onItemsRemoved={onItemsRemovedSpy}
        />,
        root
      ) as unknown) as TypedBaseUnifiedPicker;

      if (!picker.inputElement) {
        throw new Error('picker input element not set!');
      }
      picker.inputElement.value = '';
      ReactTestUtils.Simulate.keyDown(picker.inputElement, { which: KeyCodes.backspace });

      expect(onItemsRemovedSpy).toBeCalled();
      expect(picker.value).toEqual([
        {
          key: '1',
          name: 'one'
        }
      ]);

      ReactDOM.unmountComponentAtNode(root);
    });

    it('Filters out removed elements on SelectedItems.onItemsRemoved callback called', () => {
      const root = document.createElement('div');
      document.body.appendChild(root);
      const onItemsRemovedSpy = jasmine.createSpy();

      let selectedItemsonItemsRemovedCallback: ((items: ISimple[]) => void) | undefined;
      const SimpleItem = (item: ISelectedItemProps<ISimple>) => <>{item.item.name}</>;
      const InjectedSelectedItemsComponent: IUnifiedPickerProps<ISimple>['onRenderSelectedItems'] = props => {
        selectedItemsonItemsRemovedCallback = props.onItemsRemoved;
        return <SelectedItemsList<ISimple> {...props} onRenderItem={SimpleItem} />;
      };

      const items = [{ key: '1', name: 'one' }, { key: '2', name: 'two' }];

      const picker: TypedBaseUnifiedPicker = (ReactDOM.render(
        <BaseUnifiedPickerWithType
          defaultSelectedItems={items}
          onRenderFloatingPicker={BasicFloatingSuggestions}
          onItemsRemoved={onItemsRemovedSpy}
          onRenderSelectedItems={InjectedSelectedItemsComponent}
        />,
        root
      ) as unknown) as TypedBaseUnifiedPicker;

      if (!picker.inputElement) {
        throw new Error('picker input element not set!');
      }
      if (!selectedItemsonItemsRemovedCallback) {
        throw new Error('selectedItemsonItemsRemovedCallback callback not set!');
      }

      selectedItemsonItemsRemovedCallback([items[0]]);

      expect(onItemsRemovedSpy).toBeCalled();
      expect(picker.value).toEqual([
        {
          key: '2',
          name: 'two'
        }
      ]);

      ReactDOM.unmountComponentAtNode(root);
    });
  });
});
