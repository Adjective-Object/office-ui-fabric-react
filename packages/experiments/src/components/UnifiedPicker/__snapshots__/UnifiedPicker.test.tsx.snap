// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnifiedPicker renders BaseUnifiedPicker correctly 1`] = `
Array [
  <div
    className="ms-FocusZone"
    data-focuszone-id="FocusZone0"
    onCopy={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDownCapture={[Function]}
    role="presentation"
  >
    <div
      className="ms-SelectionZone"
      onClick={[Function]}
      onContextMenu={[Function]}
      onDoubleClick={[Function]}
      onFocusCapture={[Function]}
      onKeyDown={[Function]}
      onKeyDownCapture={[Function]}
      onMouseDown={[Function]}
      onMouseDownCapture={[Function]}
      role="presentation"
    >
      <div
        role="list"
      >
        <input
          aria-expanded={false}
          aria-haspopup="true"
          aria-owns="suggestion-list"
          autoCapitalize="off"
          autoComplete="off"
          data-lpignore={true}
          onChange={[Function]}
          onClick={[Function]}
          onCompositionEnd={[Function]}
          onCompositionStart={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          role="combobox"
          value=""
        />
      </div>
    </div>
  </div>,
  <div
    className="ms-BasePicker ms-BaseFloatingPicker"
  />,
]
`;
